File: dawn_go/veb_queue.go
// On RAM priority queues, Thorup '96

type VEBTree struct {
    value  int
    left   *VEBTree
    right  *VEBTree
    isLeaf bool
}

func NewVEBTree(value int) *VEBTree {
    return &VEBTree{value: value, isLeaf: true}
}

func (t *VEBTree) Min() int {
    if t.isLeaf {
        return t.value
    }
    minLeft, minRight := math.MaxInt64, math.MaxInt64
    if t.left != nil {
        minLeft = t.left.Min()
    }
    if t.right != nil {
        minRight = t.right.Min()
    }
    return min(minLeft, minRight)
}

func (t *VEBTree) Insert(value int) *VEBTree {
    if t.isLeaf {
        if value < t.value {
            t.right = NewVEBTree(t.value)
            t.value = value
        } else {
            t.right = NewVEBTree(value)
        }
        t.isLeaf = false
        return t
    }

    if value < t.left.Min() {
        t.left = t.left.Insert(value)
    } else {
        t.right = t.right.Insert(value)
    }
    return t
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: dawn_cpp/veb_queue.cpp
#include <algorithm>
#include <utility>

// On RAM priority queues, Thorup '96 
template <typename T>
struct VEBTree {
    T value;
    VEBTree* left = nullptr;
    VEBTree* right = nullptr;
    bool isLeaf;

    VEBTree(T val) : value(val), isLeaf(true) {}

    ~VEBTree() {
        delete left;
        delete right;
    }

    T min() const {
        if (isLeaf) return value;
        T minLeft = (left != nullptr) ? left->min() : std::numeric_limits<T>::max();
        T minRight = (right != nullptr) ? right->min() : std::numeric_limits<T>::max();
        return std::min(minLeft, minRight);
    }

    void insert(T val) {
        if (isLeaf) {
            if (val < value) {
                right = new VEBTree(value);
                value = val;
            } else {
                right = new VEBTree(val);
            }
            isLeaf = false;
            return;
        }

        if (val < left->min()) {
            if (left == nullptr) left = new VEBTree(val);
            else left->insert(val);
        } else {
            if (right == nullptr) right = new VEBTree(val);
            else right->insert(val);
        }
    }
};